kernel/thread.c

```diff
@@ -36,6 +36,38 @@ extern struct _static_thread_data _static_thread_data_list_end[];
 	     thread_data < _static_thread_data_list_end; \
 	     thread_data++)
 
+#if defined(CONFIG_THREAD_MONITOR)
+void k_thread_foreach(thread_user_cb user_cb, void *user_data)
+{
+	struct k_thread *thread;
+
+	__ASSERT(user_cb, "user_cb can not be NULL");
+
+	for (thread = _kernel.threads; thread != NULL;
```

Comment:
The iteration should happen while holding a lock; otherwise, if the linked list is modified midway, it'll either be called with invalid parameters, or following the next pointer will lead you to someplace you shouldn't be going.